niworkflows/interfaces/nibabel.py

```diff
@@ -586,14 +679,24 @@ def _gen_reference(
 
     # Moving images may not be RAS/LPS (more generally, transverse-longitudinal-axial)
     reoriented_moving_img = nb.as_closest_canonical(nb.load(moving_image))
-    new_zooms = reoriented_moving_img.header.get_zooms()[:3]
-
-    # Avoid small differences in reported resolution to cause changes to
-    # FOV. See https://github.com/nipreps/fmriprep/issues/512
-    # A positive diagonal affine is RAS, hence the need to reorient above.
-    new_affine = np.diag(np.round(new_zooms, 3))
 
-    resampled = nli.resample_img(fixed_image, target_affine=new_affine, interpolation='nearest')
+    if target_resolution is not None:
+        new_affine, new_shape = _calculate_target_affine(reoriented_moving_img, target_resolution)
+    else:
+        new_zooms = reoriented_moving_img.header.get_zooms()[:3]
+
+        # Avoid small differences in reported resolution to cause changes to
+        # FOV. See https://github.com/nipreps/fmriprep/issues/512
+        # A positive diagonal affine is RAS, hence the need to reorient above.
+        new_affine = np.diag(np.round(new_zooms, 3))
+        new_shape = fixed_image.shape[:3]
```

Comment:
Well, now I'm looking at this again, and we were previously enforcing a diagonal matrix, and letting nilearn figure out the translations and bounding boxes. Can we just do that again with:

```suggestion
    new_zooms = target_resolution if target_resolution is not None else reoriented_moving_img.header.get_zooms()[:3]

    # Avoid small differences in reported resolution to cause changes to
    # FOV. See https://github.com/nipreps/fmriprep/issues/512
    # A positive diagonal affine is RAS, hence the need to reorient above.
    new_affine = np.diag(np.round(new_zooms, 3))

    resampled = nli.resample_img(fixed_image, target_affine=new_affine, interpolation='nearest')
```

Apologies for bouncing around. My head is all over the place right now, so please feel free to argue back...